### DISCUSSION TITLE
cogctl communication to multiple instances does not work

### DISCUSSION THREAD

**Discussion by ventosus** [Igalia/cog#541] [Q&A] [answered]
With two cog instances (built with system DBUS):

    cog --gapplication-app-id=com.example.ONE uri
    cog --gapplication-app-id=com.example.TWO uri

There is no way to talk to both of them with cogctl by using the object path: (only the first one loaded can be talked to)

    cogctl --system --object-path=/com/example/ONE reload -> OK

    cogctl --system --object-path=/com/example/TWO reload
    GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: Object does not exist at path ?/com/example/TWO?

There is no way to talk to both of them with cogctl by using the appid:

    cogctl --system --appid=com.example.ONE reload
    GDBus.Error:org.freedesktop.DBus.Error.NameHasNoOwner: Name "com.example.ONE" does not exist

    cogctl --system --appid=com.example.TWO reload
    GDBus.Error:org.freedesktop.DBus.Error.NameHasNoOwner: Name "com.example.TWO" does not exist

How should I talk to different cog instances? Am I doing something wrong?

**Comment by ventosus** [comment]
# cog --version
    0.12.4 (WPE WebKit 2.36.3)

**Comment by chrisvollorenew** [comment]
Hello @ventosus - I also have been having trouble with cogctl and dbus. 
I wonder if theres something I am missing with my set up. 
I am running cog as root in an embedded environment. 
The system bus is at /run/dbus/...
The environment variables for dbus are not getting set "automatically" in my case. 
Have you tried checking your environment variables (such as DBUS_SYSTEM_BUS_ADDRESS) for dbus?

Please find my post here:
https://github.com/Igalia/cog/issues/532

**Comment by ventosus** [comment]
@chrisvollorenew dbus ist set up just fine on my side. my problem is addressing multiple instances of cog. addressing a single instance works just fine.

**Comment by aperezdc** [comment, marked as answer]
The same identifiers passed to `cog --gapplication-app-id=` must be passed to `cogctl --appid=`:

```sh
# Start two instances with different application identifiers
cog --gapplication-app-id=com.example.Cog1 wpewebkit.org
cog --gapplication-app-id=com.example.Cog2 igalia.com
# Load different pages on each using cogctl
cogctl --appid=com.example.Cog1 open webkit.org
cogctl --appid=com.example.Cog2 open apple.com
```
You can see this in action in the screenshot below (note that `-A` is the same as `--appid=`):

![Screenshot of a labwc compositor session, with two Cog instances and a terminal in the background running cogctl to open different pages on them](https://user-images.githubusercontent.com/723451/221872498-8e70a186-7855-453a-a35f-2b7df965ee9d.png)

**Reply by aperezdc** [reply]
Elaborating further, the error:

```
GDBus.Error:org.freedesktop.DBus.Error.NameHasNoOwner: Name "com.example.TWO" does not exist
```

means that `cogctl` has been able to connect to D-Bus (connection error has _not_ been reported), but the control interface has not been registered by Cog in _the bus accessed by `cogctl`_. Please make sure that:

- Cog can itself can access D-Bus. Unfortunately GLib doesn't warn when the control interface wasn't registered to the session bus (on the other hand, Cog warns when failing to register to the system bus&mdash;note that we recommend using the session bus regardless of this).
- Both Cog and `cogctl` are using the same bus (system bus, or session bus of the same user).

**Comment by ventosus** [comment]
Please see my initial report, I'm using the system bus. There it still does not work. Can you reproduce on the system bus, please?

    # cog --version
    0.16.1 (WPE WebKit 2.38.5)

**Comment by daydaychen** [comment]
Maybe your dbus-daemon program isn't starting? Check if this file `/run/usr/<user_id>/bus` exists

**Comment by woutervanh** [comment]
Was indeed not working on system bus, as the passed appid was not parsed yet when setting up the bus, so registering with default name com.igalia.Cog in stead of the desired name.
Fixed and created pull request: https://github.com/Igalia/cog/pull/762

**Reply by aperezdc** [reply]
Thanks! The fix was merged a few days ago.
